### PR TITLE
Refactor Torchvision Model for Classification Semi-SL

### PR DESCRIPTION
**Pull request by harimkang**
### Summary

- Add functionality for Semi-SL Task
- Add Semi-SL TV model recipes

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by chuneuny-emily — DISMISSED** [review]

File: src/otx/algo/classification/torchvision_model.py [hunk 1]
```diff
@@ -155,45 +154,148 @@ def __init__(
...
+                feature_channel = last_layer.in_features
+                head = nn.Linear(feature_channel, self.num_classes)
+
+            if self.train_type == "semi_supervised":
+                self.neck = nn.Sequential(*layers) if layers else None
+                head = OTXSemiSLLinearClsHead(
+                    num_classes=self.num_classes,
+                    in_channels=feature_channel,
+                    loss=self.loss_module,
+                )
```
Comment (sungchul1) [inline src/otx/algo/classification/torchvision_model.py]: If self.train_type == "semi_supervised", `head` seems to be assigned twice.
How about updating if-else statement?

File: src/otx/algo/classification/torchvision_model.py [hunk 2]
```diff
@@ -155,45 +154,148 @@ def __init__(
...
         mode: str = "tensor",
-    ) -> torch.Tensor:
+        **kwargs,
+    ) -> dict[str, tuple | torch.Tensor] | tuple | torch.Tensor:
         """Performs forward pass of the model.
 
         Args:
-            images (torch.Tensor): The input images.
-            labels (torch.Tensor): The ground truth labels.
+            images (torch.Tensor | dict[str, torch.Tensor]): The input images.
```
Comment (sungchul1) [inline src/otx/algo/classification/torchvision_model.py:195]: Could you simply explain the case for `dict[str, torch.Tensor]`? For semi-sl?

File: src/otx/algo/classification/torchvision_model.py [hunk 3]
```diff
@@ -155,45 +154,148 @@ def __init__(
...
+                  linear layer. Usually returns a tensor.
+
+                Defaults to "neck".
+
+        Returns:
+            dict[str, tuple | torch.Tensor] | tuple | torch.Tensor: The output of specified stage.
+            The output depends on detailed implementation. In general, the
+            output of backbone and neck is a tuple.
+        """
+        if isinstance(inputs, dict):
```
Comment (sungchul1) [inline src/otx/algo/classification/torchvision_model.py:239]: I think it should be checked if all keys for semi-sl are dict.
If not, value error can occur at L258~260.
What do you think?

File: src/otx/algo/classification/torchvision_model.py (hunk 2 above)
Comment (harimkang) [reply src/otx/algo/classification/torchvision_model.py:195]: Yes this is for Semi-SL, I added the docstring. https://github.com/openvinotoolkit/training_extensions/pull/3614/commits/6bc9bfa48aa11b0f5b4677f2e6a489e830ceda75

File: src/otx/algo/classification/torchvision_model.py (hunk 1 above)
Comment (harimkang) [reply src/otx/algo/classification/torchvision_model.py]: Thanks, Fixed - https://github.com/openvinotoolkit/training_extensions/pull/3614/commits/6bc9bfa48aa11b0f5b4677f2e6a489e830ceda75

File: src/otx/algo/classification/torchvision_model.py (hunk 3 above)
Comment (harimkang) [reply src/otx/algo/classification/torchvision_model.py:239]: Added syntax to check for key, thanks. https://github.com/openvinotoolkit/training_extensions/pull/3614/commits/6bc9bfa48aa11b0f5b4677f2e6a489e830ceda75

File: src/otx/algo/classification/torchvision_model.py [hunk 4]
```diff
@@ -155,45 +154,154 @@ def __init__(
             optimize_gap=True,
         )
 
+    def _get_head(self, net: nn.Module) -> nn.Module:
+        """Returns the head of the model."""
+        head: nn.Module
+        last_layer = list(net.children())[-1]
+        layers = []
+        if self.task == "multiclass":
```
Comment (sungchul1) [inline src/otx/algo/classification/torchvision_model.py:162]: Simple question. If `self.task != "multiclass"`, is head not created?

Comment (harimkang) [reply src/otx/algo/classification/torchvision_model.py:162]: Yes, currently only for multi-class tasks available. However, i plan to expand this (multi-label, h-label) in the future.

**Review by sungchul1 — APPROVED** [review]

**Review by chuneuny-emily — APPROVED** [review]
